text/0523-model-argument-for-route-templates.md

```diff
@@ -0,0 +1,115 @@
+`this.model`.
+
+Before:
+
+```hbs
+{{!-- The model for this route is the current user --}}
+
+<div>
+  Hi <img src="{{this.model.profileImage}}" alt="{{this.model.name}}'s profile picture"> {{this.model.name}},
+  this is a valid Ember template!
+</div>
+
+{{#if this.model.isAdmin}}
+  <div>Remember, with great power comes great responsibility!</div>
+{{/if}}
+```
+
+After:
+
+```hbs
+{{!-- The model for this route is the current user --}}
+
+<div>
+  Hi <img src="{{@model.profileImage}}" alt="{{@model.name}}'s profile picture"> {{@model.name}},
+  this is a valid Ember template!
+</div>
+
+{{#if @model.isAdmin}}
+  <div>Remember, with great power comes great responsibility!</div>
+{{/if}}
+```
+
+## Motivation
+
+With the Octane release, templates are taking on a more important role in
+idomatic Ember apps. As template become more self-sufficient, many cases where
+a JavaScript class was needed in the past (e.g. to customize the "wrapper"
+element) can now be expressed with templates alone. This is a direction we will
+continue post-Octane.
+
+We would like to update the learning materials (guides) to focus on teaching
+the template-centric component model. For example, components can be introduced
+as a way to break up large templates into smaller, named pieces, similar to
+refactoring a big function into smaller ones. Then, we can layer on making them
+reusable through passing arguments. Next, we can introduce the idea of passing
+a block and yielding. Finally, we can introduce the component class when we are
+ready to discuss adding behavior to the component.
+
+As you can see, we can accomplish quite a lot with templates-only components in
+Octane, and focusing on the teaching template first-and-foremost would provide
+a gentle learning curve for developers and designers who are comfortable with
+HTML but perhaps new to Ember (or even JavaScript). With this flow, the concept
+of a component class, and consequently, the idea of a "`this` context" in a
+template comes up quite late.
+
+This presents a problem, as we may want or need to teach route templates before
+that was introduced. As the model can only be accessed through `this.model` in
+a route template, we would be forced to introduce that concept (and the contept
+of a controller) at an earlier time than would be ideal.
+
+Providing `@model` to route templates would solve this problem quite elegantly.
+We will now be able to introduce the concept of arguments (`@name`) once and it
+can be applied consistently between component and route templates.
+
+This also aligns with the general mental model that arguments are things that
+are passed into the template from the outside (which is true in the case of the
+route model).
+
+This can also be thought of as a small incremental step in the bigger picture
+of reforming route templates and removing controllers from Ember. Specifically,
+it moves us a bit closer to the mental model that controllers/route templates
```

Comment:
I'm curious how this might interact with any future proposals for routes being able to import components, etc.? Does this in any way fence off paths in that direction or make them more difficult? And likewise if we added some *other* kind of magic hooks to routes as a few RFCs have proposed—we obviously don't need to design those here, but this has *implications* for any such RFC in that it would *very strongly* suggest similar synthesis of arguments.

Reply:
I think the general idea is that routes should just invoke a component. For example, we may evolve in a direction where routes just have a `args` hook that resolves into the arguments that should be passed into the component. In which case they would definitely want to be accessible from the template using the named argument syntax. In that world, the model hook and `@model` could be a special case/default behavior for the more generalized `args` hook. That is a transition we need to design and plan out anyway, but I _do_ think we would want "things that are passed in from the route" to be treated as component arguments in any possible future.

Reply:
> the model hook and @model could be a special case/default behavior for the more generalized args hook.

I think this is the right general direction, but we should consider skipping this interim step, especially since eliminating controllers is on the 2019-2020 roadmap.

I have concerns about further codifying anything related to the `model` / `beforeModel` / `afterModel` hooks. Going straight to a more generalized `args` hook (or something similar) would allow someone to define `@model` as an arg if they want without any possible confusion/conflict with the model hooks. And it would allow us to move forward to a controller-free future by using the presence of an `args` hook to avoid creating an associated controller entirely (while instead allowing for an optional backing component class).

Reply:
I don't think there is a conflict. Let's say we go with that design/migration path. The presence of a `model` hook implies the `@model` argument. If you have opted into some future `args` thing then it would be up to you to decide what argument(s) to pass to your component.

```js
class UserRoute {
  model(params) {
    return this.store.find(params.userId);
  }
}
```

...desugars into...

```js
class UserRoute {
  async args(params) {
    let model = await this.store.find(params.userId);
    return { model };
  }
}
```

Something like that.

Reply:
@chancancode I agree that's the logical migration path if we go ahead with this.

The question in my mind is whether this approach is worth the temporary incoherence:

* It allows access to both `this.model` and `@model` in the same template (and allows them to diverge).

* Bridging an "old" and "new" world means that it is less clear from many angles which world you are working in. Looking at a route template with `@model` means that you won't know whether you have a backing controller or not. You also may not realize that model hooks are in use in any given route instead of `args`.

An alternative would be to pursue the `args` approach immediately and allow developers to make a clean break with models and controllers when they are ready. The presence of the args hook could cause a hard error if it coexists with any model hooks. Furthermore, it would be a signal that no controller should be created.

For me the question comes down to whether we can build consensus around an `args`-like approach quickly. If not, then I would be in favor of merging this as a Plan B. Although I only focused on the drawbacks, I also recognize the benefits.